File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        
        unordered_map<Node*, Node*> hash;

        Node* cur = head; // copy all nodes
        while (cur)
        {
            Node* copy = new Node(cur->val);
            hash[cur] = copy;
            cur = cur->next;
        }
        cur = head; // set pointers in all nodes
        while (cur)
        {
            Node* copy = hash[cur];
            copy->next = hash[cur->next];
            copy->random = hash[cur->random];
            cur = cur->next;

        }
        
        return hash[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;
    
    public Node(int _val) {
        val = _val;
        next = null;
        random = null;
    }
}
*/

public class Solution {
    public Node CopyRandomList(Node head)
    {
        var hash = new Dictionary<Node, Node>();

        Node cur = head;
        while (cur != null)
        {
            Node copy = new Node(cur.val);
            hash[cur] = copy;
            cur = cur.next;
        }

        cur = head;
        while (cur != null) 
        {
            Node copy = hash[cur];
            copy.next = cur.next != null ? hash[cur.next] : null;
            copy.random = cur.random != null ? hash[cur.random] : null;
            cur = cur.next;
        }
        
        return head != null ? hash[head] : null;
    }
}
